File: src/inheritbench/capability/evaluator.py
"""Declarative structured-output evaluator and safety-rule interpreter."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Literal, cast

from jsonschema import Draft202012Validator

from inheritbench.artifacts.hashing import canonical_json, content_sha256
from inheritbench.capability.json_pointer import pointer_exists, resolve_pointer
from inheritbench.capability.plugins import TrustedEvaluatorPlugin
from inheritbench.capability.schemas import (
    CapabilityInputRecord,
    CapabilityOracleRecord,
    EvaluatorConfig,
    GenericEvaluationResult,
    ParserFinding,
    SafetyFinding,
)

_FENCE = re.compile(r"\A\s*```(?:json)?\s*\n(?P<body>.*)\n```\s*\Z", re.DOTALL | re.IGNORECASE)
_MISSING = object()


def evaluate_output(
    *,
    record: CapabilityInputRecord,
    oracle: CapabilityOracleRecord,
    raw_output: str,
    config: EvaluatorConfig,
    output_schema: dict[str, Any],
    cross_field_schema: dict[str, Any],
    vocabularies: dict[str, set[Any]],
    safety_rules: list[dict[str, Any]],
    trusted_plugin: TrustedEvaluatorPlugin | None = None,
) -> GenericEvaluationResult:
    if trusted_plugin is not None:
        result = trusted_plugin.evaluate(
            record_id=record.record_id,
            raw_output=raw_output,
            model_input=record.payload,
            oracle=oracle.model_dump(mode="json"),
        )
        if result.record_id != record.record_id:
            raise ValueError("trusted evaluator returned the wrong record ID")
        return GenericEvaluationResult.model_validate(result, strict=True)
    strict_candidate = raw_output.strip()
    parsed, strict_findings = _parse_candidate(strict_candidate)
    normalized_candidate: str | None = None
    classification = "UNPARSEABLE"
    findings = strict_findings
    if parsed is not None:
        classification = "STRICT_VALID"
    elif config.whole_output_json_fence:
        match = _FENCE.fullmatch(raw_output)
        if match is not None:
            normalized_candidate = match.group("body").strip()
            parsed, findings = _parse_candidate(normalized_candidate)
            if parsed is not None:
                classification = "NORMALIZED_VALID"

    schema_valid = False
    vocabulary_conformant = False
    cross_field_conformant = False
    if parsed is not None:
        errors = sorted(Draft202012Validator(output_schema).iter_errors(parsed), key=str)
        if errors:
            findings = [
                ParserFinding(code="SCHEMA_INVALID", message=error.message) for error in errors
            ]
        else:
            schema_valid = True
            vocabulary_conformant = not _vocabulary_errors(parsed, config, vocabularies)
            cross_field_conformant = not list(
                Draft202012Validator(cross_field_schema).iter_errors(parsed)
            )

    field_correctness: dict[str, bool] = {}
    evaluation_eligible = (
        schema_valid
        and parsed is not None
        and (cross_field_conformant or not config.strict_requires_cross_field)
    )
    if evaluation_eligible and parsed is not None:
        for rule in config.comparisons:
            predicted = resolve_pointer(parsed, rule.pointer, _MISSING)
            expected = resolve_pointer(oracle.expected, rule.pointer, _MISSING)
            field_correctness[rule.name] = _compare(predicted, expected, rule.mode, rule.tolerance)
    else:
        field_correctness = {rule.name: False for rule in config.comparisons}

    semantic_names = {rule.name for rule in config.comparisons if rule.semantic}
    semantic_match = (
        evaluation_eligible
        and bool(semantic_names)
        and all(field_correctness.get(name, False) for name in semantic_names)
    )
    structural_exact = evaluation_eligible and parsed == oracle.expected
    historical_strict_valid = (
        classification == "STRICT_VALID"
        and schema_valid
        and (cross_field_conformant or not config.strict_requires_cross_field)
    )
    safety_document = {
        "prediction": parsed,
        "expected": oracle.expected,
        "input": record.payload,
        "oracle": oracle.safety_context,
        "metrics": {
            "semantic_match": semantic_match,
            "structural_exact": structural_exact,
            "schema_valid": schema_valid,
        },
    }
    safety_findings = (
        _evaluate_safety_rules(safety_document, safety_rules) if parsed is not None else []
    )
    mean = sum(field_correctness.values()) / len(field_correctness) if field_correctness else 0.0
    payload = {
        "schema_version": "inheritbench.generic-evaluation.v0.2",
        "record_id": record.record_id,
        "raw_output": raw_output,
        "strict_candidate": strict_candidate,
        "normalized_candidate": normalized_candidate,
        "parser_classification": classification,
        "parse_valid": parsed is not None,
        "valid_json": parsed is not None,
        "schema_valid": schema_valid,
        "vocabulary_conformant": vocabulary_conformant,
        "cross_field_conformant": cross_field_conformant,
        "historical_strict_valid": historical_strict_valid,
        "strict_valid": historical_strict_valid,
        "structural_exact": structural_exact,
        "semantic_match": semantic_match,
        "field_correctness": field_correctness,
        "mean_field_correctness": mean,
        "parsed_output": parsed,
        "expected": oracle.expected,
        "parser_findings": findings,
        "safety_findings": safety_findings,
        "coverage": oracle.coverage,
    }
    payload["content_sha256"] = content_sha256(payload)
    return GenericEvaluationResult.model_validate(payload, strict=True)


def load_vocabularies(root: Path, config: EvaluatorConfig) -> dict[str, set[Any]]:
    values: dict[str, set[Any]] = {}
    for pointer, relative in config.controlled_vocabularies.items():
        raw = json.loads((root / relative).read_text(encoding="utf-8"))
        if not isinstance(raw, list):
            raise ValueError(f"vocabulary {relative} must contain a JSON list")
        values[pointer] = {_hashable(item) for item in raw}
    return values


def _parse_candidate(candidate: str) -> tuple[dict[str, Any] | None, list[ParserFinding]]:
    if not candidate:
        return None, [ParserFinding(code="INVALID_JSON", message="output is empty")]
    try:
        value = json.loads(candidate)
    except json.JSONDecodeError as exc:
        return None, [ParserFinding(code="INVALID_JSON", message=str(exc))]
    if not isinstance(value, dict):
        return None, [ParserFinding(code="ROOT_NOT_OBJECT", message="JSON root must be an object")]
    return value, []


def _vocabulary_errors(
    parsed: dict[str, Any],
    config: EvaluatorConfig,
    vocabularies: dict[str, set[Any]],
) -> list[ParserFinding]:
    findings: list[ParserFinding] = []
    for pointer in config.required_pointers:
        if not pointer_exists(parsed, pointer):
            findings.append(
                ParserFinding(code="SCHEMA_INVALID", message=f"missing required pointer {pointer}")
            )
    for pointer, allowed in vocabularies.items():
        value = resolve_pointer(parsed, pointer, _MISSING)
        if value is _MISSING or _hashable(value) not in allowed:
            findings.append(
                ParserFinding(
                    code="SCHEMA_INVALID",
                    message=f"value at {pointer} is not in the controlled vocabulary",
                )
            )
    return findings


def _compare(predicted: Any, expected: Any, mode: str, tolerance: float | None) -> bool:
    if predicted is _MISSING or expected is _MISSING:
        return False
    if mode in {"exact", "list"}:
        return bool(predicted == expected)
    if mode == "set":
        if not isinstance(predicted, list) or not isinstance(expected, list):
            return False
        return {_hashable(value) for value in predicted} == {_hashable(value) for value in expected}
    if mode == "numeric":
        if isinstance(predicted, bool) or isinstance(expected, bool):
            return False
        if not isinstance(predicted, int | float) or not isinstance(expected, int | float):
            return False
        return abs(float(predicted) - float(expected)) <= (tolerance or 0.0)
    raise ValueError(f"unsupported comparison mode {mode}")


def _hashable(value: Any) -> Any:
    if isinstance(value, dict | list):
        return canonical_json(value)
    return value


def _evaluate_safety_rules(
    document: dict[str, Any], rules: list[dict[str, Any]]
) -> list[SafetyFinding]:
    findings: list[SafetyFinding] = []
    for rule in rules:
        if _eval_ast(rule["when"], document):
            findings.append(
                SafetyFinding(
                    code=str(rule["code"]),
                    severity=cast(
                        Literal["info", "warning", "blocker"],
                        str(rule["severity"]),
                    ),
                    message=str(rule["message"]),
                )
            )
    return findings


def validate_safety_ast(node: Any) -> None:
    if not isinstance(node, dict) or len(node) != 1:
        raise ValueError("safety expression must be a one-key object")
    operator, value = next(iter(node.items()))
    if operator in {"and", "or"}:
        if not isinstance(value, list) or not value:
            raise ValueError(f"{operator} requires a non-empty list")
        for child in value:
            validate_safety_ast(child)
        return
    if operator == "not":
        validate_safety_ast(value)
        return
    if operator in {"exists", "missing"}:
        if not isinstance(value, str) or not value.startswith("/"):
            raise ValueError(f"{operator} requires a JSON Pointer")
        return
    if operator in {"eq", "ne", "in", "not_in", "contains"}:
        if not isinstance(value, dict) or set(value) != {"pointer", "value"}:
            raise ValueError(f"{operator} requires pointer and value")
        if not isinstance(value["pointer"], str) or not value["pointer"].startswith("/"):
            raise ValueError(f"{operator} pointer is invalid")
        expected = value["value"]
        if isinstance(expected, dict):
            if set(expected) != {"pointer"}:
                raise ValueError(f"{operator} value reference must contain only pointer")
            if not isinstance(expected["pointer"], str) or not expected["pointer"].startswith("/"):
                raise ValueError(f"{operator} value reference pointer is invalid")
        return
    raise ValueError(f"unsupported safety operator {operator}")


def _eval_ast(node: dict[str, Any], document: dict[str, Any]) -> bool:
    operator, value = next(iter(node.items()))
    if operator == "and":
        return all(_eval_ast(child, document) for child in value)
    if operator == "or":
        return any(_eval_ast(child, document) for child in value)
    if operator == "not":
        return not _eval_ast(value, document)
    if operator == "exists":
        return pointer_exists(document, value)
    if operator == "missing":
        return not pointer_exists(document, value)
    actual = resolve_pointer(document, value["pointer"], _MISSING)
    expected_spec = value["value"]
    expected = (
        resolve_pointer(document, expected_spec["pointer"], _MISSING)
        if isinstance(expected_spec, dict) and set(expected_spec) == {"pointer"}
        else expected_spec
    )
    if operator == "eq":
        return bool(actual == expected)
    if operator == "ne":
        return bool(actual != expected)
    if operator == "in":
        return actual in expected if isinstance(expected, list) else False
    if operator == "not_in":
        return actual not in expected if isinstance(expected, list) else False
    if operator == "contains":
        return expected in actual if isinstance(actual, list | str | dict) else False
    raise ValueError(operator)

File: apps/web/src/lib/sandbox/evaluator.ts
import Ajv2020 from "ajv/dist/2020.js";

import { canonicalJson, contentSha256 } from "./hashing";
import {
  evaluationResultSchema,
  type EvaluationContract,
  type EvaluationResult,
  type JsonValue,
  type Prediction,
  type RecordDefinition,
} from "./schemas";

const fence = /^\s*```(?:json)?\s*\n([\s\S]*)\n```\s*$/i;
const missing = Symbol("missing");
type Missing = typeof missing;

function tokens(pointer: string): string[] {
  if (pointer === "") return [];
  if (!pointer.startsWith("/")) throw new Error(`invalid JSON Pointer: ${pointer}`);
  return pointer
    .slice(1)
    .split("/")
    .map((part) => part.replace(/~1/g, "/").replace(/~0/g, "~"));
}

export function resolvePointer(document: unknown, pointer: string, fallback: unknown = missing): unknown {
  let current: unknown = document;
  for (const token of tokens(pointer)) {
    if (current === null || typeof current !== "object" || !(token in current)) return fallback;
    current = (current as Record<string, unknown>)[token];
  }
  return current;
}

export function pointerExists(document: unknown, pointer: string): boolean {
  return resolvePointer(document, pointer, missing) !== missing;
}

function parseCandidate(candidate: string): {
  parsed: Record<string, JsonValue> | null;
  findings: EvaluationResult["parser_findings"];
} {
  if (!candidate) {
    return { parsed: null, findings: [{ code: "INVALID_JSON", message: "output is empty" }] };
  }
  let value: unknown;
  try {
    value = JSON.parse(candidate);
  } catch (error) {
    return {
      parsed: null,
      findings: [{ code: "INVALID_JSON", message: error instanceof Error ? error.message : String(error) }],
    };
  }
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return {
      parsed: null,
      findings: [{ code: "ROOT_NOT_OBJECT", message: "JSON root must be an object" }],
    };
  }
  return { parsed: value as Record<string, JsonValue>, findings: [] };
}

function deepEqual(left: unknown, right: unknown): boolean {
  if (Object.is(left, right)) return true;
  if (Array.isArray(left) && Array.isArray(right)) {
    return left.length === right.length && left.every((value, index) => deepEqual(value, right[index]));
  }
  if (
    left !== null &&
    right !== null &&
    typeof left === "object" &&
    typeof right === "object" &&
    !Array.isArray(left) &&
    !Array.isArray(right)
  ) {
    const leftKeys = Object.keys(left).sort();
    const rightKeys = Object.keys(right).sort();
    return (
      deepEqual(leftKeys, rightKeys) &&
      leftKeys.every((key) =>
        deepEqual((left as Record<string, unknown>)[key], (right as Record<string, unknown>)[key]),
      )
    );
  }
  return false;
}

function hashable(value: unknown): string | number | boolean | null | undefined {
  return value !== null && typeof value === "object"
    ? canonicalJson(value as JsonValue)
    : (value as string | number | boolean | null | undefined);
}

function compare(
  predicted: unknown | Missing,
  expected: unknown | Missing,
  mode: "exact" | "list" | "set" | "numeric",
  tolerance?: number | null,
): boolean {
  if (predicted === missing || expected === missing) return false;
  if (mode === "exact" || mode === "list") return deepEqual(predicted, expected);
  if (mode === "set") {
    if (!Array.isArray(predicted) || !Array.isArray(expected)) return false;
    const left = new Set(predicted.map(hashable));
    const right = new Set(expected.map(hashable));
    return left.size === right.size && [...left].every((value) => right.has(value));
  }
  return (
    typeof predicted === "number" &&
    typeof expected === "number" &&
    Math.abs(predicted - expected) <= (tolerance ?? 0)
  );
}

export function validateSafetyAst(node: unknown): void {
  if (node === null || typeof node !== "object" || Array.isArray(node) || Object.keys(node).length !== 1) {
    throw new Error("safety expression must be a one-key object");
  }
  const [operator, value] = Object.entries(node)[0];
  if (operator === "and" || operator === "or") {
    if (!Array.isArray(value) || value.length === 0) throw new Error(`${operator} requires a non-empty list`);
    value.forEach(validateSafetyAst);
    return;
  }
  if (operator === "not") {
    validateSafetyAst(value);
    return;
  }
  if (operator === "exists" || operator === "missing") {
    if (typeof value !== "string" || !value.startsWith("/")) throw new Error(`${operator} requires a JSON Pointer`);
    return;
  }
  if (["eq", "ne", "in", "not_in", "contains"].includes(operator)) {
    if (
      value === null ||
      typeof value !== "object" ||
      Array.isArray(value) ||
      Object.keys(value).sort().join(",") !== "pointer,value"
    ) {
      throw new Error(`${operator} requires pointer and value`);
    }
    const operand = value as Record<string, unknown>;
    if (typeof operand.pointer !== "string" || !operand.pointer.startsWith("/")) {
      throw new Error(`${operator} pointer is invalid`);
    }
    if (operand.value !== null && typeof operand.value === "object" && !Array.isArray(operand.value)) {
      const reference = operand.value as Record<string, unknown>;
      if (
        Object.keys(reference).join(",") !== "pointer" ||
        typeof reference.pointer !== "string" ||
        !reference.pointer.startsWith("/")
      ) {
        throw new Error(`${operator} value reference is invalid`);
      }
    }
    return;
  }
  throw new Error(`unsupported safety operator ${operator}`);
}

function evalAst(node: Record<string, JsonValue>, document: Record<string, JsonValue>): boolean {
  const [operator, value] = Object.entries(node)[0];
  if (operator === "and") return (value as JsonValue[]).every((child) => evalAst(child as Record<string, JsonValue>, document));
  if (operator === "or") return (value as JsonValue[]).some((child) => evalAst(child as Record<string, JsonValue>, document));
  if (operator === "not") return !evalAst(value as Record<string, JsonValue>, document);
  if (operator === "exists") return pointerExists(document, value as string);
  if (operator === "missing") return !pointerExists(document, value as string);
  const operand = value as Record<string, JsonValue>;
  const actual = resolvePointer(document, operand.pointer as string, missing);
  const expectedSpec = operand.value;
  const expected =
    expectedSpec !== null &&
    typeof expectedSpec === "object" &&
    !Array.isArray(expectedSpec) &&
    Object.keys(expectedSpec).length === 1 &&
    "pointer" in expectedSpec
      ? resolvePointer(document, expectedSpec.pointer as string, missing)
      : expectedSpec;
  if (operator === "eq") return deepEqual(actual, expected);
  if (operator === "ne") return !deepEqual(actual, expected);
  if (operator === "in") return Array.isArray(expected) && expected.some((item) => deepEqual(actual, item));
  if (operator === "not_in") return Array.isArray(expected) && !expected.some((item) => deepEqual(actual, item));
  if (operator === "contains") {
    if (Array.isArray(actual)) return actual.some((item) => deepEqual(item, expected));
    if (typeof actual === "string") return typeof expected === "string" && actual.includes(expected);
    return actual !== null && typeof actual === "object" && typeof expected === "string" && expected in actual;
  }
  throw new Error(operator);
}

export async function evaluateOutput(args: {
  record: RecordDefinition;
  prediction: Prediction;
  contract: EvaluationContract;
}): Promise<EvaluationResult> {
  const { record, prediction, contract } = args;
  const config = contract.evaluator;
  const strictCandidate = prediction.raw_output.trim();
  let { parsed, findings } = parseCandidate(strictCandidate);
  let normalizedCandidate: string | null = null;
  let classification: EvaluationResult["parser_classification"] = "UNPARSEABLE";
  if (parsed !== null) classification = "STRICT_VALID";
  else if (config.whole_output_json_fence) {
    const match = fence.exec(prediction.raw_output);
    if (match) {
      normalizedCandidate = match[1].trim();
      ({ parsed, findings } = parseCandidate(normalizedCandidate));
      if (parsed !== null) classification = "NORMALIZED_VALID";
    }
  }

  const ajv = new Ajv2020({ allErrors: true, strict: false });
  let schemaValid = false;
  let vocabularyConformant = false;
  let crossFieldConformant = false;
  if (parsed !== null) {
    const outputValidator = ajv.compile(contract.schemas.output);
    const outputValid = outputValidator(parsed);
    if (!outputValid) {
      findings = (outputValidator.errors ?? [])
        .map((error) => ({ code: "SCHEMA_INVALID" as const, message: error.message ?? "schema invalid" }))
        .sort((a, b) => a.message.localeCompare(b.message));
    } else {
      schemaValid = true;
      vocabularyConformant =
        config.required_pointers.every((pointer) => pointerExists(parsed, pointer)) &&
        Object.entries(config.controlled_vocabularies).every(([pointer, relative]) => {
          const name = relative.split("/").at(-1)?.replace(/\.json$/, "") ?? relative;
          const allowed = contract.vocabularies[name] ?? [];
          const value = resolvePointer(parsed, pointer, missing);
          return value !== missing && allowed.some((item) => deepEqual(value, item));
        });
      const crossFieldValidator = ajv.compile(contract.schemas.cross_field);
      crossFieldConformant = crossFieldValidator(parsed);
    }
  }

  const eligible = schemaValid && parsed !== null && (crossFieldConformant || !config.strict_requires_cross_field);
  const fieldCorrectness = Object.fromEntries(
    config.comparisons.map((rule) => [
      rule.name,
      eligible
        ? compare(
            resolvePointer(parsed, rule.pointer, missing) as unknown | Missing,
            resolvePointer(record.expected, rule.pointer, missing) as unknown | Missing,
            rule.mode,
            rule.tolerance,
          )
        : false,
    ]),
  );
  const semanticNames = config.comparisons.filter((rule) => rule.semantic).map((rule) => rule.name);
  const semanticMatch =
    eligible && semanticNames.length > 0 && semanticNames.every((name) => fieldCorrectness[name]);
  const structuralExact = eligible && deepEqual(parsed, record.expected);
  const historicalStrictValid =
    classification === "STRICT_VALID" &&
    schemaValid &&
    (crossFieldConformant || !config.strict_requires_cross_field);
  const safetyDocument = {
    prediction: parsed,
    expected: record.expected,
    input: record.input,
    oracle: record.safety_context,
    metrics: { semantic_match: semanticMatch, structural_exact: structuralExact, schema_valid: schemaValid },
  } as Record<string, JsonValue>;
  const safetyFindings =
    parsed === null
      ? []
      : contract.safety.rules
          .filter((rule) => {
            validateSafetyAst(rule.when);
            return evalAst(rule.when, safetyDocument);
          })
          .map(({ code, severity, message }) => ({ code, severity, message }));
  const values = Object.values(fieldCorrectness);
  const payload = {
    schema_version: "inheritbench.generic-evaluation.v0.2" as const,
    record_id: record.record_id,
    raw_output: prediction.raw_output,
    strict_candidate: strictCandidate,
    normalized_candidate: normalizedCandidate,
    parser_classification: classification,
    parse_valid: parsed !== null,
    valid_json: parsed !== null,
    schema_valid: schemaValid,
    vocabulary_conformant: vocabularyConformant,
    cross_field_conformant: crossFieldConformant,
    historical_strict_valid: historicalStrictValid,
    strict_valid: historicalStrictValid,
    structural_exact: structuralExact,
    semantic_match: semanticMatch,
    field_correctness: fieldCorrectness,
    mean_field_correctness: values.length ? values.filter(Boolean).length / values.length : 0,
    parsed_output: parsed,
    expected: record.expected,
    parser_findings: findings,
    safety_findings: safetyFindings,
    coverage: record.coverage,
  };
  return evaluationResultSchema.parse({
    ...payload,
    content_sha256: await contentSha256(payload as unknown as JsonValue),
  });
}
